src/parsing/staticallyCompileXPath.ts

```diff
@@ -44,69 +52,153 @@ export default function staticallyCompileXPath(
-			const moduleURI = moduleImports[modulePrefix];
-			enhanceStaticContextWithModule(rootStaticContext, moduleURI);
+	return compileAstToExpression(queryBodyContents, compilationOptions);
+}
 
-			rootStaticContext.registerNamespace(modulePrefix, moduleURI);
-		});
+export default function staticallyCompileXPath(
+	selector: EvaluableExpression,
+	compilationOptions: {
+		allowUpdating: boolean | undefined;
+		allowXQuery: boolean | undefined;
+		annotateAst: boolean | undefined;
+		debug: boolean | undefined;
+		disableCache: boolean | undefined;
+	},
+	namespaceResolver: (namespace: string) => string | null,
+	variables: { [varName: string]: any },
+	moduleImports: { [namespaceURI: string]: string },
+	defaultFunctionNamespaceURI: string,
+	functionNameResolver: FunctionNameResolver
+): { expression: Expression; staticContext: StaticContext } {
+	const executionSpecificStaticContext = new ExecutionSpecificStaticContext(
+		namespaceResolver,
+		variables,
+		defaultFunctionNamespaceURI,
+		functionNameResolver
+	);
+	const rootStaticContext = new StaticContext(executionSpecificStaticContext);
+	Object.keys(moduleImports).forEach((modulePrefix) => {
+		const moduleURI = moduleImports[modulePrefix];
+		enhanceStaticContextWithModule(rootStaticContext, moduleURI);
 
-		expression.performStaticEvaluation(rootStaticContext);
+		rootStaticContext.registerNamespace(modulePrefix, moduleURI);
+	});
 
-		if (!compilationOptions.disableCache) {
-			storeStaticCompilationResultInCache(
-				xpathString,
-				language,
-				executionSpecificStaticContext,
-				moduleImports,
-				expression,
-				compilationOptions.debug,
-				defaultFunctionNamespaceURI
-			);
+	if (typeof selector === 'string') {
+		selector = normalizeEndOfLines(selector);
+		const result = createExpressionFromString(
+			selector,
+			compilationOptions,
+			namespaceResolver,
+			variables,
+			moduleImports,
+			defaultFunctionNamespaceURI,
+			functionNameResolver
+		);
+		switch (result.state) {
+			case CACHE_STATE.STATIC_ANALYZED:
+				return {
+					staticContext: rootStaticContext,
+					expression: result.expression,
+				};
+
+			case CACHE_STATE.COMPILED:
+				result.expression.performStaticEvaluation(rootStaticContext);
+
+				return {
+					staticContext: rootStaticContext,
+					expression: result.expression,
+				};
+			case CACHE_STATE.PARSED:
+				const expressionFromAst = buildExpressionFromAst(
+					result.ast,
+					compilationOptions,
+					namespaceResolver,
+					variables,
+					moduleImports,
+					defaultFunctionNamespaceURI,
+					functionNameResolver,
+					rootStaticContext
+				);
+				expressionFromAst.performStaticEvaluation(rootStaticContext);
+
+				return {
+					staticContext: rootStaticContext,
+					expression: expressionFromAst,
+				};
 		}
 	}
 
-	return { expression, staticContext: rootStaticContext };
+	// AST is an element: convert to jsonml
+	const ast = convertXmlToAst(selector as Element);
```

Comment:
This cast should not be needed I think

Reply:
now evaluableExpression is either string or Element, so this cast is removed